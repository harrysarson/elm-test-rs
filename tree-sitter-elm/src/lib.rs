extern "C" {
    fn tree_sitter_elm() -> tree_sitter::Language;
}

pub fn language() -> tree_sitter::Language {
    unsafe { tree_sitter_elm() }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_display_snapshot;
    use std::fs;
    use std::path::PathBuf;
    use tree_sitter::Parser;

    #[test]
    fn test_language() {
        let mut parser = Parser::new();
        parser.set_language(language()).unwrap();

        let source_file: PathBuf = ["c-src", "examples", "basic.elm"].iter().collect();
        let source_code = fs::read_to_string(source_file).unwrap();

        let tree = parser.parse(source_code, None).unwrap();
        let root_node = tree.root_node();

        let sexp = sexp::parse(&root_node.to_sexp()).expect("valid sexp");

        assert_display_snapshot!(sexp.pretty());
    }
}
