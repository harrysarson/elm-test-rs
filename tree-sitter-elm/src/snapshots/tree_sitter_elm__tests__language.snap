---
source: tree-sitter-elm/src/lib.rs
expression: sexp.pretty()
---
(file
  (module_declaration
    (module)
    (upper_case_qid
      (upper_case_identifier)
    )
    (exposing_list
      (exposing)
      (left_parenthesis)
      (exposed_type
        (upper_case_identifier)
        (exposed_union_constructors
          (left_parenthesis)
          (double_dot)
          (right_parenthesis)
        )
      )
      (comma)
      (exposed_value
        (lower_case_identifier)
      )
      (comma)
      (exposed_value
        (lower_case_identifier)
      )
      (comma)
      (exposed_value
        (lower_case_identifier)
      )
      (right_parenthesis)
    )
  )
  (import_clause
    (import)
    (upper_case_qid
      (upper_case_identifier)
    )
  )
  (import_clause
    (import)
    (upper_case_qid
      (upper_case_identifier)
    )
    (exposing_list
      (exposing)
      (left_parenthesis)
      (exposed_type
        (upper_case_identifier)
      )
      (comma)
      (exposed_value
        (lower_case_identifier)
      )
      (comma)
      (exposed_value
        (lower_case_identifier)
      )
      (comma)
      (exposed_value
        (lower_case_identifier)
      )
      (right_parenthesis)
    )
  )
  (import_clause
    (import)
    (upper_case_qid
      (upper_case_identifier)
      (dot)
      (upper_case_identifier)
    )
    (exposing_list
      (exposing)
      (left_parenthesis)
      (exposed_value
        (lower_case_identifier)
      )
      (right_parenthesis)
    )
  )
  (value_declaration
    (function_declaration_left
      (lower_case_identifier)
    )
    (eq)
    (function_call_expr
      (value_expr
        (value_qid
          (upper_case_identifier)
          (dot)
          (lower_case_identifier)
        )
      )
      (record_expr
        (left_brace)
        (field
          (lower_case_identifier)
          (eq)
          (number_constant_expr
            (number_literal)
          )
        )
        (comma)
        (field
          (lower_case_identifier)
          (eq)
          (value_expr
            (value_qid
              (lower_case_identifier)
            )
          )
        )
        (comma)
        (field
          (lower_case_identifier)
          (eq)
          (value_expr
            (value_qid
              (lower_case_identifier)
            )
          )
        )
        (right_brace)
      )
    )
  )
  (type_declaration
    (type)
    (upper_case_identifier)
    (eq)
    (union_variant
      (upper_case_identifier)
    )
    (pipe)
    (union_variant
      (upper_case_identifier)
    )
  )
  (value_declaration
    (function_declaration_left
      (lower_case_identifier)
      (lower_pattern
        (lower_case_identifier)
      )
      (lower_pattern
        (lower_case_identifier)
      )
    )
    (eq)
    (case_of_expr
      (case)
      (value_expr
        (value_qid
          (lower_case_identifier)
        )
      )
      (of)
      (case_of_branch
        (pattern
          (union_pattern
            (upper_case_qid
              (upper_case_identifier)
            )
          )
        )
        (arrow)
        (bin_op_expr
          (value_expr
            (value_qid
              (lower_case_identifier)
            )
          )
          (operator
            (operator_identifier)
          )
          (number_constant_expr
            (number_literal)
          )
        )
      )
      (case_of_branch
        (pattern
          (union_pattern
            (upper_case_qid
              (upper_case_identifier)
            )
          )
        )
        (arrow)
        (bin_op_expr
          (value_expr
            (value_qid
              (lower_case_identifier)
            )
          )
          (operator
            (operator_identifier)
          )
          (number_constant_expr
            (number_literal)
          )
        )
      )
    )
  )
  (value_declaration
    (function_declaration_left
      (lower_case_identifier)
      (lower_pattern
        (lower_case_identifier)
      )
    )
    (eq)
    (function_call_expr
      (value_expr
        (value_qid
          (lower_case_identifier)
        )
      )
      (list_expr
        (left_square_bracket)
        (right_square_bracket)
      )
      (list_expr
        (left_square_bracket)
        (function_call_expr
          (value_expr
            (value_qid
              (lower_case_identifier)
            )
          )
          (list_expr
            (left_square_bracket)
            (function_call_expr
              (value_expr
                (value_qid
                  (lower_case_identifier)
                )
              )
              (value_expr
                (upper_case_qid
                  (upper_case_identifier)
                )
              )
            )
            (right_square_bracket)
          )
          (list_expr
            (left_square_bracket)
            (function_call_expr
              (value_expr
                (value_qid
                  (lower_case_identifier)
                )
              )
              (string_constant_expr
                (open_quote)
                (regular_string_part)
                (close_quote)
              )
            )
            (right_square_bracket)
          )
        )
        (comma)
        (function_call_expr
          (value_expr
            (value_qid
              (lower_case_identifier)
            )
          )
          (list_expr
            (left_square_bracket)
            (right_square_bracket)
          )
          (list_expr
            (left_square_bracket)
            (function_call_expr
              (value_expr
                (value_qid
                  (lower_case_identifier)
                )
              )
              (parenthesized_expr
                (left_parenthesis)
                (function_call_expr
                  (value_expr
                    (value_qid
                      (upper_case_identifier)
                      (dot)
                      (lower_case_identifier)
                    )
                  )
                  (value_expr
                    (value_qid
                      (lower_case_identifier)
                    )
                  )
                )
                (right_parenthesis)
              )
            )
            (right_square_bracket)
          )
        )
        (comma)
        (function_call_expr
          (value_expr
            (value_qid
              (lower_case_identifier)
            )
          )
          (list_expr
            (left_square_bracket)
            (function_call_expr
              (value_expr
                (value_qid
                  (lower_case_identifier)
                )
              )
              (value_expr
                (upper_case_qid
                  (upper_case_identifier)
                )
              )
            )
            (right_square_bracket)
          )
          (list_expr
            (left_square_bracket)
            (function_call_expr
              (value_expr
                (value_qid
                  (lower_case_identifier)
                )
              )
              (string_constant_expr
                (open_quote)
                (regular_string_part)
                (close_quote)
              )
            )
            (right_square_bracket)
          )
        )
        (right_square_bracket)
      )
    )
  )
)
